kratos/factories/base_factory.h

```diff
@@ -0,0 +1,342 @@
+    {
+        KRATOS_ERROR << "Methods must be implemented in the base class" << std::endl;
+        return false;
+    }
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const
+    {
+        return "BaseFactoryMethods";
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream& rOStream) const
+    {
+        rOStream << this->Info() << std::endl;
+    }
+
+    /// Print object's data.
+    virtual void PrintData(std::ostream& rOStream) const
+    {
+//         rOStream << this->Info() << std::endl;
+    }
+};
+
+/**
+ * @class BaseFactory
+ * @ingroup KratosCore
+ * @brief Here we add the functions needed for the registration of class
+ * @details Defines the base class factory
+ * @author Vicente Mataix Ferrandiz
+ * @tparam TClass The class to create the factory
+ * @tparam TAuxiliarClass The auxiliar class to create the factory
+ */
+template<typename TClass, typename TAuxiliarClass = TClass>
+class BaseFactory
+    : public BaseFactoryMethods
+{
+public:
+    ///@name Type Definitions
+    ///@{
+        
+    /// The definition of the class
+    typedef TClass ClassType;
+
+    /// The definition of the auxiliar class
+    typedef TAuxiliarClass AuxiliarClassType;
+
+    /// Pointer definition of BaseFactory
+    KRATOS_CLASS_POINTER_DEFINITION(BaseFactory);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /** Constructor.
+     */
+    explicit BaseFactory(){}
+
+    /** Destructor.
+     */
+    virtual ~BaseFactory(){}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method checks if the linear class is registered
+     * @param rClassName The nanme of the class
+     * @return True if registered, false otherwise
+     */
+    bool Has(const std::string& rClassName) const override
+    {
+        return KratosComponents<ClassType>::Has( rClassName );
+    }
+
+    /**
+     * @brief This method creates a new class
+     * @param rModel The model containing the problem
+     * @param Settings The settings of the factory
+     * @return The pointer to the class of interest
+     */
+    template<typename = std::enable_if<std::is_same<TAuxiliarClass, TClass>::value>>
+    typename ClassType::Pointer Create(Kratos::Parameters Settings) const
+    {
+        const std::string& r_name = Settings["name"].GetString();
+        KRATOS_ERROR_IF_NOT(Has( r_name )) << "Trying to construct a class with type name= " << r_name << std::endl <<
+                                            "Which does not exist. The list of available options (for currently loaded applications) are: " << std::endl <<
+                                            KratosComponents<ClassType>() << std::endl;
+        const ClassType& aux = KratosComponents<ClassType>::Get( r_name );
+        return aux.Create(Settings);
+    }
```

Comment:
I am sure that I am missing some details here but a variadic template shouldn't work here:

```c++
template< typename... TArgumentsType >

    typename ClassType::Pointer Create(TArgumentsType&&... Arguments) const
    {
        const std::string& r_name = Settings["name"].GetString();
        KRATOS_ERROR_IF_NOT(Has( r_name )) << "Trying to construct a class with type name= " << r_name << std::endl <<
                                            "Which does not exist. The list of available options (for currently loaded applications) are: " << std::endl <<
                                            KratosComponents<ClassType>() << std::endl;
        const ClassType& aux = KratosComponents<ClassType>::Get( r_name );
        return aux.Create(std::forward<TArgumentsType>(Arguments);
    }
	
```
Substituting all variances of this method enabled if?

Reply:
Finally it works (with some tricks). Certantly this will easier once we move to C++17, or even C++14

Reply:
So what are the tricks?

Reply:
Create a tuple of the arguments and get always the last argument assuming it corresponds with the Parameters, use std::get and other stuff as can be seen here:

~~~cpp
template<typename... TArgumentsType >
    typename ClassType::Pointer Create(TArgumentsType&&... Arguments) const
    {
        std::tuple<TArgumentsType...> args(Arguments...);
        constexpr std::size_t args_size = std::tuple_size<std::tuple<TArgumentsType...>>::value;
        Kratos::Parameters Settings = std::get<args_size - 1>(args);
        const std::string& r_name = Settings["name"].GetString();
        KRATOS_ERROR_IF_NOT(Has( r_name )) << "Trying to construct a class with type name= " << r_name << "\n" <<
                                            "Which does not exist. The list of available options (for currently loaded applications) are: \n" <<
                                            KratosComponents<ClassType>() << std::endl;
        const ClassType& aux = KratosComponents<ClassType>::Get( r_name );
        return aux.Create(std::forward<TArgumentsType>(Arguments)...);
    }
~~~

Reply:
@pooyan-dadvand see the previous comment